zboxcore/sdk/downloadworker.go

```
@@ -46,338 +45,494 @@ type DownloadRequest struct {
+	suite := edwards25519.NewBlakeSHA256Ed25519()
+	reEncMessage := &encryption.ReEncryptedMessage{
+		D1: suite.Point(),
+		D4: suite.Point(),
+		D5: suite.Point(),
+	}
+	err = reEncMessage.Unmarshal(result.BlockChunks[blockNum])
 	if err != nil {
-		if req.statusCallback != nil {
-			logger.Logger.Error(err.Error())
-			req.statusCallback.Error(req.allocationID, remotePathCallback, OpDownload, errors.Wrap(err, "Can't create local file"))
-		}
-		return
+		logger.Logger.Error("ReEncrypted Block unmarshall failed", req.blobbers[result.idx].Baseurl, err)
+		return nil, err
 	}
-	defer wrFile.Close()
-	req.isDownloadCanceled = false
-	if req.statusCallback != nil {
-		req.statusCallback.Started(req.allocationID, remotePathCallback, OpDownload, int(size))
+	decrypted, err := req.encScheme.ReDecrypt(reEncMessage)
+	if err != nil {
+		logger.Logger.Error("Block redecryption failed", req.blobbers[result.idx].Baseurl, err)
+		return nil, err
 	}
+	return decrypted, nil
+}
 
-	if req.endBlock == 0 {
-		req.endBlock = chunksPerShard
+// processDownload will setup download parameters and downloads data with given
+// start block, end block and number of blocks to download in single request.
+// This will also write data to the file and will verify content by calculating content hash.
+func (req *DownloadRequest) processDownload(ctx context.Context) {
+	if req.completedCallback != nil {
+		defer req.completedCallback(req.remotefilepath, req.remotefilepathhash)
 	}
 
-	if req.startBlock >= req.endBlock {
-		if req.statusCallback != nil {
-			sys.Files.Remove(req.localpath)
-			req.statusCallback.Error(req.allocationID, remotePathCallback, OpDownload, errors.New("invalid_block_num", "start block should be less than end block"))
-		}
+	remotePathCB := req.remotefilepath
+	if remotePathCB == "" {
+		remotePathCB = req.remotefilepathhash
+	}
+	fRef, err := req.getFileRef(remotePathCB)
+	if err != nil {
+		logger.Logger.Error(err.Error())
```

We should return error to caller instead of just logging. The caller may rely on the result to decide whether it should continue or not.